### ISSUE TITLE
DrawParam matrices are incorrect

### ISSUE THREAD

**Issue by PSteinhaus** [ggez/ggez#879]
While implementing my BitmapFont ontop of SpriteBatch I stumbled upon this line in ```drawparam.rs```:
```
/// A [`DrawParam`](struct.DrawParam.html) that has been crunched down to a single matrix.
    fn to_na_matrix(&self) -> Matrix4 {
        // ...
        translate * offset * rotation * scale * offset_inverse     // this one
    }
```
And it baffled me, because it seemed wrong to me (though I use DrawParams in my own game and they seem to work just fine).
The problem is this:

The matrix should be calculated as:
```translate * rotation * scale * offset_inverse```

because this is exactly what we want to do:
- shift the sprite by an offset (called offset_inverse in the code)
- then scale it around that point
- rotate it around that point
- and finally move it to its destination

There simply is no need for multiplying offset onto the matrix after rotation.

Now this seems like a major problem, right? But why has noone ever noticed it, if it really is?
The reason is that most sprites are big and drawn inside of a large coordinate system, thereby making a shift of say 0.5 pixels practically unnoticable.

But if you use a smaller sprite and a small coordinate system as in the following example:
```
use ggez;

use ggez::event;
use ggez::graphics;
use ggez::{Context, GameResult};
use std::env;
use std::path;
use ggez::nalgebra::{Point2};
use ggez::graphics::{DrawParam, Image, Rect, FilterMode};

struct MainState {
    img: Image
}

impl MainState {
    fn new(ctx: &mut Context) -> GameResult<MainState> {
        let s = MainState { img: Image::new(ctx, "/testSprite.png").unwrap() };
        Ok(s)
    }
}

impl event::EventHandler for MainState {
    fn update(&mut self, _ctx: &mut Context) -> GameResult {
        Ok(())
    }

    fn draw(&mut self, ctx: &mut Context) -> GameResult {
        graphics::clear(ctx, [0.1, 0.2, 0.3, 1.0].into());

        let offset = Point2::new(0.5, 0.5);

        graphics::draw(ctx, &self.img, DrawParam::default().offset(offset))?;
        graphics::present(ctx)?;

        Ok(())
    }
}

pub fn main() -> GameResult {
    let resource_dir = if let Ok(manifest_dir) = env::var("CARGO_MANIFEST_DIR") {
        let mut path = path::PathBuf::from(manifest_dir);
        path.push("resources");
        path
    } else {
        path::PathBuf::from("./resources")
    };

    let cb = ggez::ContextBuilder::new("helloworld", "ggez").add_resource_path(resource_dir);
    let (ctx, event_loop) = &mut cb.build()?;
    graphics::set_screen_coordinates(ctx, Rect::new(0.0, 0.0, 4.0, 4.0))?;
    graphics::set_default_filter(ctx, FilterMode::Nearest);

    let state = &mut MainState::new(ctx)?;
    event::run(ctx, event_loop, state)
}
```
where the testSprite is in my case only 2x2 pixels in size, then you get the following behavior:

With ```offset = Point2::new(0.0, 0.0);``` everything works as expected:
![offset0](https://user-images.githubusercontent.com/40324631/105479958-6774d000-5ca5-11eb-9628-17d1532f437e.png)

With ```offset = Point2::new(0.5, 0.5);``` things start to go south:
![offset0_5](https://user-images.githubusercontent.com/40324631/105480014-7fe4ea80-5ca5-11eb-801d-030c1e970370.png)
We would expect to only see the green square, but this is not the case since the sprite is moved contrary to the offset after it is scaled.

Another good example is ```offset = Point2::new(0.9, 0.9);```:
![offset0_9](https://user-images.githubusercontent.com/40324631/105480215-bde20e80-5ca5-11eb-9aa0-454ce4cff033.png)
Here we would expect to see close to nothing of the sprite, but instead we get the image we would expect from ```offset = Point2::new(0.45, 0.45);```.
(This is because the ```real_scale``` that this sprite will get assigned is 2.0).

I'll test whether my hypothesis is correct and simply removing ```offset``` from the multiplication fixes this in a minute (or a few hours depending on my work) and start a pull request if true.

**Comment by PSteinhaus** [comment]
Forking ggez I just noticed I'm a bit behind current developements. As of now the matrix seems to be calculated a lot differently (by hand that is).
But I just checked and the behavior is still the same.

**Comment by PSteinhaus** [comment]
Solved in #880.
